services/src/main/java/org/keycloak/services/DefaultComponentFactoryProviderFactory.java

```diff
@@ -0,0 +1,191 @@
+import org.keycloak.component.ComponentModel;
+import org.keycloak.component.ComponentModelScope;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.provider.InvalidationHandler;
+import org.keycloak.provider.InvalidationHandler.InvalidableObjectType;
+import org.keycloak.provider.InvalidationHandler.ObjectType;
+import org.keycloak.provider.Provider;
+import org.keycloak.provider.ProviderFactory;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.jboss.logging.Logger;
+
+/**
+ * @author hmlnarik
+ */
+public class DefaultComponentFactoryProviderFactory implements ComponentFactoryProviderFactory {
+
+    private static final Logger LOG = Logger.getLogger(DefaultComponentFactoryProviderFactory.class);
+    public static final String PROVIDER_ID = "default";
+
+    private final AtomicReference<ConcurrentMap<String, ProviderFactory>> componentsMap = new AtomicReference<>(new ConcurrentHashMap<>());
+
+    /**
+     * Should an ID in the key be invalidated, it would invalidate also all the IDs in the values
+     */
+    private final ConcurrentMap<Object, Set<String>> dependentInvalidations = new ConcurrentHashMap<>();
+
+    private KeycloakSessionFactory factory;
+    private boolean componentCachingAvailable;
+
+    @Override
+    public void init(Scope config) {
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        this.factory = factory;
+        this.componentCachingAvailable = this.factory.getProviderFactory(ClusterProvider.class) != null;
+        if (! componentCachingAvailable) {
+            LOG.warn("No system-wide ClusterProviderFactory found. Cannot send messages across cluster, thus disabling caching of components.");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Provider> ProviderFactory<T> getProviderFactory(Class<T> clazz, String realmId, String componentId, Function<KeycloakSessionFactory, ComponentModel> modelGetter) {
+        ProviderFactory res = componentsMap.get().get(componentId);
+        if (res != null) {
+            LOG.tracef("Found cached ProviderFactory for %s in (%s, %s)", clazz, realmId, componentId);
+            return res;
+        }
+
+        // Apply the expensive operation before putting it into the cache
+        final ComponentModel cm;
+        if (modelGetter == null) {
+            LOG.debugf("Getting component configuration for component (%s, %s) from realm configuration", clazz, realmId, componentId);
+            AtomicReference<ComponentModel> cmr = new AtomicReference<>();
```

Comment:
This can be probably simplified a bit to call `KeycloakModelUtils.componentModelGetter` as this code is in fact duplicate of it.